pubtools/_quay/remove_image.py

```diff
@@ -0,0 +1,323 @@
+
+    return sorted(list(set(references_to_remove)))
+
+
+# TODO: integration tests
+def remove_images(
+    reference,
+    quay_api_token,
+    quay_user,
+    quay_password,
+    pyxis_server,
+    pyxis_krb_principal,
+    pyxis_krb_ktfile,
+    send_umb_msg=False,
+    umb_urls=[],
+    umb_cert=None,
+    umb_client_key=None,
+    umb_ca_cert=None,
+    umb_topic="VirtualTopic.eng.pub.quay_remove_images",
+):
+    """
+    Remove quay images. All tags referencing a given image will be removed.
+
+    The purpose of this entrypoint is to get image to the state where it will be garbage
+    collected in the future (no tag references it). V2S2 manifests and manifest lists will be
+    treated as separate categories. In other words, partial ML removal is not supported, and
+    single-arch images will not be  removed if also a part of multiarch image.
+
+    Args:
+        reference (str):
+            Images specified by tags to remove. Comma separated values.
+        quay_api_token (str):
+            OAuth token for authentication of Quay REST API.
+        quay_user (str):
+            Quay username for Docker HTTP API.
+        quay_password (str):
+            Quay password for Docker HTTP API.
+        pyxis_server (str):
+            Pyxis service hostname:
+        pyxis_krb_principal (str):
+            Pyxis kerberos principal in form: name@REALM.
+        pyxis_krb_ktfile (str):
+            Pyxis Kerberos client keytab.
+        send_umb_msg (bool):
+            Whether to send UMB messages about the untagged images.
+        umb_urls ([str]):
+            AMQP broker URLs to connect to.
+        umb_cert (str):
+            Path to a certificate used for UMB authentication.
+        umb_client_key (str):
+            Path to a client key to decrypt the certificate (if necessary).
+        umb_ca_cert (str):
+            Path to a CA certificate (for mutual authentication).
+        umb_topic (str):
+            Topic to send the UMB messages to.
+    """
+    verify_remove_image_args(reference, send_umb_msg, umb_urls, umb_cert)
+    quay_client = QuayClient(quay_user, quay_password)
+
+    references_to_remove = []
+    repo_images_mapping = group_images_by_repo(reference)
+
+    for references in sorted(repo_images_mapping.values()):
+        references_to_remove += get_repo_images_to_remove(references, quay_client)
+
+    references_to_remove = sorted(list(set(references_to_remove)))
+    untag_images(
+        references_to_remove,
+        quay_api_token=quay_api_token,
+        remove_last=True,
+        quay_user=quay_user,
+        quay_password=quay_password,
+        send_umb_msg=True,
+        umb_urls=umb_urls,
+        umb_cert=umb_cert,
+        umb_client_key=umb_client_key,
+        umb_ca_cert=umb_ca_cert,
+    )
+
+    sig_remover = SignatureRemover()
+    sig_remover.set_quay_client(quay_client)
+    for reference in references_to_remove:
+        sig_remover.remove_tag_signatures(
```

Comment:
It can't remove signatures of already untagged images.

Reply:
Good point, changed the order.